[5.2] pagination presenters return Htmlable object

harrygr opened the pull request:
Make the pagination contracts return an instance of
`Illuminate\View\Expression` so that safe blade tags can be used.

Same as #10878 but for the 5.2 branch

Commit:
pagination presenters return view expressions

Make the pagination contracts return an instance of
`Illuminate\View\Expression` so that safe blade tags can be used.

GrahamCampbell commented:
I think this was rejected before because it adds an extra dependency on the view component to the paginator, and one could argue the paginator shouldn't know or care about the view component.

taylorotwell commented:
Yeah will have to add view dependency to paginator.

harrygr commented:
Alternatively we could just make the Expression optional in the contract and still allow an implementation to return a string.

Commit:
added view dependency, made expression optional

An implementation of `Illuminate\Contracts\PaginationPresenter` may
return a string instead of an Expression.

GrahamCampbell commented:
> Alternatively we could just make the Expression optional in the contract and still allow an implementation to return a string.

rkgrep commented:
Why don't use `Htmlable` which is available in `contracts`?

harrygr commented:
@rkgrep that wouldn't remove the dependency on view, and `htmlable` does not have a `__toString()` method so not all implementations can be directly echoed.

rkgrep commented:
https://github.com/laravel/framework/blob/5.1/src/Illuminate/Support/HtmlString.php

harrygr commented:
That looks like the answer. I'll update my PR

Commit:
Use HtmlString instead of view Expression

Removes dependency on View as HtmlString is already in Support

rkgrep commented on src/Illuminate/Pagination/SimpleBootstrapThreePresenter.php:
```diff
@@ -30,16 +31,16 @@ public function hasPages()
      * Convert the URL window into Bootstrap HTML.
      *
-     * @return string
+     * @return \Illuminate\View\Expression
```
`\Illuminate\Support\HtmlString`

rkgrep commented:
Paginator [supports](https://github.com/laravel/framework/blob/5.1/src/Illuminate/Pagination/AbstractPaginator.php#L483) `__toString()`, so it should implement `Htmlable` either - as people may use `{{$paginator}}`
